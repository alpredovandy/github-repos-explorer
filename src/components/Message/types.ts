export interface MessageProps {
    show: boolean;
    onClose: () => void;
    delay?: number;
    message: string;
    type: 'success' | 'error' | 'info';
}
