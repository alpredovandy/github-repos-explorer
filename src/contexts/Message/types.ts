export interface MessageData {
    type?: 'success' | 'error' | 'info';
    message?: string;
}

export interface MessageState {
    /** open message */
    show: boolean;

    /** message data */
    data?: MessageData;
}

export interface MessageContextProps {
    showMessage: (props: MessageData) => void;
    showSuccessMessage: (props: Omit<MessageData, 'type'>) => void;
    showErrorMessage: (props: Omit<MessageData, 'type'>) => void;
    closeMessage: () => void;
}
