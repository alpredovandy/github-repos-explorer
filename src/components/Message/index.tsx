import { FC, useEffect } from 'react';

import { MessageProps } from '@/components/Message/types';

const Message: FC<MessageProps> = ({ show, onClose, delay = 3000, message, type }) => {
    const alertType = type === 'success' ? 'alert-success' : 'alert-error';

    useEffect(() => {
        if (delay) {
            setTimeout(() => {
                onClose();
            }, delay);
        }
    }, [show, delay, onClose]);

    if (!show) return null;

    return (
        <div className="toast toast-top toast-end">
            <div className={`alert ${alertType ?? 'alert-info'}`}>
                <span>{message}</span>
            </div>
        </div>
    );
};

export default Message;
