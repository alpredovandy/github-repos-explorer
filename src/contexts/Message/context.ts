import { createContext, useContext } from 'react';

import type { MessageContextProps } from './types';

export const MessageContext = createContext<MessageContextProps>({} as MessageContextProps);

export const useMessageContext = () => {
    const value = useContext(MessageContext);

    return value;
};
