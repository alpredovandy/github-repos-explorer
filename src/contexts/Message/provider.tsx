import React, { useCallback, useMemo, useState } from 'react';

import Message from '@/components/Message';

import { MessageContext } from './context';
import type { MessageContextProps, MessageState } from './types';

const initialMessageState: MessageState = {
    show: false,
    data: {},
};

const MessageProvider = ({ children }: { children: React.ReactNode }) => {
    const [message, setMessage] = useState(initialMessageState);

    const showMessage: MessageContextProps['showMessage'] = useCallback((data) => {
        setMessage({
            show: true,
            data,
        });
    }, []);

    const showSuccessMessage: MessageContextProps['showSuccessMessage'] = useCallback(
        (data) => {
            showMessage({
                type: 'success',
                ...data,
            });
        },
        [showMessage]
    );

    const showErrorMessage: MessageContextProps['showErrorMessage'] = useCallback(
        (data) => {
            showMessage({
                type: 'error',
                ...data,
            });
        },
        [showMessage]
    );

    const closeMessage = useCallback(() => {
        setMessage((prev) => ({ ...prev, show: false }));
    }, []);

    const memoizeValue: MessageContextProps = useMemo(
        () => ({
            showMessage,
            showSuccessMessage,
            showErrorMessage,
            closeMessage,
        }),
        [showMessage, showSuccessMessage, showErrorMessage, closeMessage]
    );

    return (
        <MessageContext.Provider value={memoizeValue}>
            {children}
            <Message
                onClose={closeMessage}
                show={message.show}
                delay={3000}
                type={message.data?.type ?? 'info'}
                message={message.data?.message ?? ''}
            />
        </MessageContext.Provider>
    );
};

export default MessageProvider;
