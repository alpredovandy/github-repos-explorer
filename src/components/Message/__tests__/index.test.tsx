import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import Message from '@/components/Message';
import type { MessageProps } from '@/components/Message/types';

describe('Test components: Message', () => {
    let props: MessageProps;
    const renderMessage = () => render(<Message {...props} />);

    it('Snapshot', () => {
        const { container } = renderMessage();

        expect(container).toMatchSnapshot();
    });
});
