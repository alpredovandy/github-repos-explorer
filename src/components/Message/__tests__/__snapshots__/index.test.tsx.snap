// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test components: Message > Snapshot 1`] = `<div />`;
